src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java

```diff
@@ -1013,4 +1013,21 @@ private boolean userHasPermissions(StudentParticipation participation) {
     public List<StudentParticipation> findWithSubmissionsWithResultByStudentIdAndExercise(Long studentId, Set<Exercise> exercises) {
         return studentParticipationRepository.findByStudentIdAndExerciseWithEagerSubmissionsResult(studentId, exercises);
     }
+
+    /**
+     * Get a mapping of participation ids to the number of submission for each participation.
+     *
+     * @param exerciseId the id of the exercise for which to consider participations
+     * @return the number of submissions per participation in the given exercise
+     */
+    public Map<Long, Integer> countSubmissionsPerParticipationByExerciseId(Long exerciseId) {
+        List<long[]> submissionCountResults = studentParticipationRepository.countSubmissionsPerParticipationByExerciseId(exerciseId);
+        Map<Long, Integer> map = new HashMap<Long, Integer>();
+        submissionCountResults.forEach(submissionCountResult -> {
+            Long participationId = submissionCountResult[0];
+            Integer submissionCount = Math.toIntExact(submissionCountResult[1]);
+            map.put(participationId, submissionCount);
+        });
```

Comment:
I think this can be easier using `.stream(toMap({}))`.

Reply:
Thanks, that does make it much shorter!